🚀 2단계 - 모델링 하기

deepredk opened the pull request:
안녕하세요 리뷰어님! 2단계 미션 수행하였습니다.
리뷰어님께 피드백 받고 그 내용 바탕으로 페어미션 진행하고 재리뷰요청 드릴 것 같습니다!

deepredk committed:
docs: 모델링 작성

Hyeon9mak commented on README.md:
```diff
@@ -209,3 +209,120 @@ docker compose -p kitchenpos up -d
+### Product
+#### 속성
+- **상품 이름 (Product Name)**: 0자 이상의 비속어가 포함되지 않은 문자열
+- **상품 가격 (Product Price)**: 0 이상의 숫자
```
모델링 <-> 코드 간 변환은 언제나 자연스러워야한다고 생각하는데요,
때문에 `비속어`와 같은 용어도 미리 영문표기를 준비해둔다면
모델링 문서만 읽고도 빠르게 코드를 작성할 수 있게 될거 같아요 😉

Hyeon9mak commented on README.md:
```diff
@@ -209,3 +209,120 @@ docker compose -p kitchenpos up -d
+  - 상품 이름은 0자 이상이어야 하며, 비속어를 포함하지 않아야 한다.
+  - 상품 가격은 0 이상이어야 한다.
+- **상품 가격 변경 (Change Product Price)**
+  - 변경된 가격으로 인해 해당 상품을 포함하는 '메뉴 구성 상품 가격의 총합'이 기존 메뉴 가격보다 낮아지면 해당 메뉴를 비공개한다.
```
😉 👍 
그런데... 생명주기를 생각해보면 무조건 상품이 먼저 탄생하고, 그 이후 메뉴가 탄생할텐데요,
상품 입장에서는 메뉴가 어떻게 달라지는지 꼭 이해하고 있어야할까요? 🤔 
반대로 메뉴가 상품을 안다면..?

deepredk replied:
아 요건 도메인 이벤트? 배치? 고런 걸로 접근이 돼야할 것 같네요
감사합니다 ㅎㅎ

Hyeon9mak commented on README.md:
```diff
@@ -209,3 +209,120 @@ docker compose -p kitchenpos up -d
+  - 메뉴 구성 상품이 1개 이상이어야 한다.
+  - 메뉴 이름은 0자 이상이어야 하며, 비속어를 포함하지 않아야 한다.
+  - 메뉴 가격은 0 이상이어야 하며,'메뉴 구성 상품 가격의 총합'보다 같거나 작아야 한다.
+- **메뉴 가격 변경 (Change Menu Price)**
```
각 행동들을 영어로 잘 준비해주신 덕분에... 바로 메서드 명으로 활용해볼 수 있겠군요 😄

Hyeon9mak commented on README.md:
```diff
@@ -209,3 +209,120 @@ docker compose -p kitchenpos up -d
+  - 메뉴 가격이 '메뉴 구성 상품 가격의 총합'보다 같거나 작아야 한다.
+- **메뉴 비공개 (Hide Menu)**
+
+#### Menu Group
+##### 속성
+- **메뉴 그룹 이름 (Menu Group Name)**: 1자 이상의 문자열
+
+##### 행동
+- **메뉴 그룹 등록 (Create Menu Group)**
+  - 메뉴 그룹 이름은 1자 이상이어야 한다.
```
행위와 속성을 분리해주신 덕분에 명확하군요 😄 👍

Hyeon9mak commented on README.md:
```diff
@@ -209,3 +209,120 @@ docker compose -p kitchenpos up -d
+  - '음식 제공됨' 상태여야 한다.
+  - 상태를 '완료'로 변경한다.
+
+#### Order Table
+##### 행동
+- **손님 수 변경 (Change Number of Guests)**
+  - 점유중이어야 한다.
+  - 변경할 손님 수가 0이상이어야 한다.
+- **테이블 점유 (Occupied)**
+  - 점유중이 아니어야 한다.
+- **테이블 미점유 (Not Occupied)**
+  - 점유중이어야 한다.
+  - 인원수를 0으로 변경한다.
+  - 미점유 상태로 변경한다.
```
OrderTable 을 매장 주문으로 포함하여주셨군요! 👍

Hyeon9mak commented on README.md:
```diff
@@ -209,3 +209,120 @@ docker compose -p kitchenpos up -d
+  - 메뉴 가격은 0 이상이어야 하며,'메뉴 구성 상품 가격의 총합'보다 같거나 작아야 한다.
+- **메뉴 가격 변경 (Change Menu Price)**
+  - 변경할 가격은 '메뉴 구성 상품 가격의 총합'보다 같거나 작아야 한다.
+- **메뉴 공개 (Display Menu)**
+  - 메뉴 가격이 '메뉴 구성 상품 가격의 총합'보다 같거나 작아야 한다.
+- **메뉴 비공개 (Hide Menu)**
```
메뉴가 공개/비공개 되는 것은 클라이언트에서 노출하지 않음을 뜻하는 것일까요, 실제 상태를 변경한다는 것일까요?

deepredk replied:
좀 더 명확하게 적어봤습니다!

Hyeon9mak commented on README.md:
```diff
@@ -209,3 +209,120 @@ docker compose -p kitchenpos up -d
+- **배달 주소 (Delivery Address)**: 1자 이상의 문자열
+- **배달 주문 상태 (Delivery Order Status)**: '대기', '수락', '음식 제공됨', '배달 시작', '배달 완료', '완료'
+
+#### 행동
+- **배달 주문 등록 (Create Delivery Order)**
+  - 배달 주소를 가져야 한다.
+  - 상태가 '대기'가 된다.
+- **배달 주문 수락 (Accept Delivery Order)**
+  - '대기' 상태여야 한다.
+  - Delivery Agency로 배달을 요청한다.
+  - 상태를 '수락'으로 변경한다.
+- **배달 음식 제공됨 (Serve Delivery Order)**
+  - '수락' 상태여야 한다.
+  - 상태를 '음식 제공됨'으로 변경한다.
+- **배달 시작 (Start Delivery)**
+  - '음식 제공됨' 상태여야 한다.
+  - 상태를 '배달 시작'으로 변경한다.
+- **배달 완료 (Complete Delivery)**
+  - '배달 시작' 상태여야 한다.
+  - 상태를 '배달 완료'로 변경한다.
+- **배달 주문 완료 (Complete Delivery Order)**
+  - '배달 완료' 상태여야 한다.
```
각 상태별로 어떤 상태로 이관될 수 있는지 명확히 정리해주셨군요.
이 덕분에 "오.. 상태패턴을 써볼까?" 같은 설계도 고민할 수 있겠네요!

Hyeon9mak reviewed (changes requested):
안녕하세요 진홍님, 모델링 잘 진행해주셨습니다! 👍 👍 
몇 가지 코멘트 남겼으니 확인해주세요 😄

deepredk committed:
docs: 리뷰어님 피드백 반영

- 헷갈릴 수 있는 용어들 영문 병기
- 상품 Bounded Context에서 메뉴 Bounded Context를 모르도록
  - 메뉴의 '정책'으로 추가하였음
- 메뉴 공개/비공개에 대해 설명 추가

deepredk commented:
페어미션에서 모델링 관련해서 크게 이견이 없었어서 리뷰어님 피드백만 반영하게 됐습니다!

Hyeon9mak commented on README.md:
```diff
@@ -209,3 +209,125 @@ docker compose -p kitchenpos up -d
+  - 메뉴 공개 상태를 '공개'로 변경한다.
+- **메뉴 비공개 (Hide Menu)**
+  - 메뉴 공개 상태를 '비공개'로 변경한다.
+
+#### 정책
+- 상품 가격 변경으로 인해 해당 상품을 포함하는 메뉴들의 '메뉴 구성 상품 가격의 총합'이 기존 메뉴 가격보다 낮아지면 해당 메뉴를 비공개해야 한다.
```
👍

Hyeon9mak reviewed (approved):
안녕하세요 진홍님! 피드백 잘 반영해주셨습니다 😄 
다음 단계로 넘어가시죠!
